renderdoc/driver/vulkan/vk_serialise.cpp

```diff
@@ -11451,7 +11470,399 @@ void Deserialise(const VkPerformanceQuerySubmitInfoKHR &el)
   DeserialiseNext(el.pNext);
 }
 
+template <typename SerialiserType>
+void DoSerialise(SerialiserType &ser, VkAabbPositionsKHR &el)
+{
+  SERIALISE_MEMBER(minX);
+  SERIALISE_MEMBER(minY);
+  SERIALISE_MEMBER(minZ);
+  SERIALISE_MEMBER(maxX);
+  SERIALISE_MEMBER(maxY);
+  SERIALISE_MEMBER(maxZ);
+}
+
+template <typename SerialiserType>
+void DoSerialise(SerialiserType &ser, VkAccelerationStructureBuildGeometryInfoKHR &el)
+{
+  RDCASSERT(ser.IsReading() ||
+            el.sType == VK_STRUCTURE_TYPE_ACCELERATION_STRUCTURE_BUILD_GEOMETRY_INFO_KHR);
+  SerialiseNext(ser, el.sType, el.pNext);
+
+  SERIALISE_MEMBER(type).Important();
+  SERIALISE_MEMBER_VKFLAGS(VkBuildAccelerationStructureFlagsKHR, flags);
+  SERIALISE_MEMBER(mode);
+  SERIALISE_MEMBER(srcAccelerationStructure);
+  SERIALISE_MEMBER(dstAccelerationStructure);
+  SERIALISE_MEMBER(geometryCount);
+
+  // flatten the indirect array into single pGeometries-like list.  Only one of ppGeometries or
+  // pGeometries can be NULL
+  if(ser.IsWriting() && el.ppGeometries)
+  {
+    VkAccelerationStructureGeometryKHR *pGeometries =
+        new VkAccelerationStructureGeometryKHR[el.geometryCount];
+    for(uint32_t i = 0; i < el.geometryCount; ++i)
+    {
+      pGeometries[i] = *(el.ppGeometries[i]);
+    }
+
+    ser.Serialise("pGeometries"_lit, pGeometries, el.geometryCount, SerialiserFlags::NoFlags);
+  }
+  else
+  {
+    SERIALISE_MEMBER_ARRAY(pGeometries, geometryCount);
```

Comment:
The general rule for serialising is that we shouldn't have multiple different serialise calls for the same object/data. The main exception to this being the initial contents, although even in that case we generally try and follow it.

If you need more complex processing like this you should do any write-time pre-processing, then serialise in a common place, then do any read-time post-processing. But the main serialise call should be bidirectional.

Reply:
Like this?
```
  VkAccelerationStructureGeometryKHR *pGeometries =
      (VkAccelerationStructureGeometryKHR *)el.pGeometries;
  if(ser.IsWriting() && el.ppGeometries)
  {
    pGeometries = new VkAccelerationStructureGeometryKHR[el.geometryCount];
    for(uint32_t i = 0; i < el.geometryCount; ++i)
    {
      pGeometries[i] = *(el.ppGeometries[i]);
    }
  }

  ser.Serialise("pGeometries"_lit, pGeometries, el.geometryCount, SerialiserFlags::AllocateMemory);

  if(ser.IsWriting() && el.ppGeometries)
    delete[] pGeometries;
  if(ser.IsReading())
  {
    el.pGeometries = pGeometries;
    el.ppGeometries = NULL;
  }
```

Reply:
yeh that looks good to me.